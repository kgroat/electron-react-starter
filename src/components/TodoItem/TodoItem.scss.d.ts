export const todoItem: string;
export const title: string;
export const checkboxIcon: string;
export const deleteIcon: string;
