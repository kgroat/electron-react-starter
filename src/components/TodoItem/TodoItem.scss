
@import '../../globalStyles/mixins';

.todoItem {
  @include itemBox();
  display: flex;

  .title {
    flex-grow: 1;
  }

  .checkboxIcon {
    @include icon(inherit, $hoverColor);
  }

  .deleteIcon {
    @include icon($errorText, $errorColor);
  }
}
