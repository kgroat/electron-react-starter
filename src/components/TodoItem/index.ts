
export { default, Props } from './TodoItem'
