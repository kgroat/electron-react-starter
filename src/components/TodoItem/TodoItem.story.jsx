
import * as React from 'react'

import { storiesOf } from '@storybook/react'

import TodoItem from './'

storiesOf('TodoItem', module)
  .add('basic example', () => <TodoItem />)
