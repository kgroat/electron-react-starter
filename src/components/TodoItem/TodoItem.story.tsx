
import * as React from 'react'

import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import TodoItem from './'

const deleteAction = action('deleted')

const checkAction = action('checked')

class Checkable extends React.Component<{}, { checked: boolean }> {
  state = {
    checked: false
  }

  toggleChecked = (checked = !this.state.checked) => {
    this.setState({ checked })
    checkAction(checked)
  }

  render () {
    return <TodoItem name='Get milk' checked={this.state.checked} onChange={this.toggleChecked} onDelete={deleteAction} />
  }
}

storiesOf('TodoItem', module)
  .add('unchecked', () => <TodoItem name='Get milk' onChange={checkAction} />)
  .add('checked', () => <TodoItem name='Get milk' checked onChange={checkAction} />)
  .add('withDelete', () => <TodoItem name='Get milk' onDelete={deleteAction} onChange={checkAction} />)
  .add('interactible', () => <Checkable />)
  .add('larger', () => <span style={{ fontSize: '2em' }}><TodoItem name='Get milk' onDelete={deleteAction} /></span>)
