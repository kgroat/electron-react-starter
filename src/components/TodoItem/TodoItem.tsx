
import * as React from 'react'

import { todoItem, title, checkboxIcon as checkboxIconClass, deleteIcon } from './TodoItem.scss'

export interface Props {
  name: string
  checked?: boolean
  onChange?: (checked: boolean) => void
  onDelete?: () => void
}

export default ({ name, checked, onChange, onDelete }: Props) => {
  const checkboxIcon = checked ? 'checkbox-marked-outline' : 'checkbox-blank-outline'

  return (
    <div className={todoItem}>
      <span className={title}>{name}</span>
      { onDelete ? <i onClick={() => onDelete()} className={`mdi mdi-delete ${deleteIcon}`} /> : null }
      <i onClick={() => onChange(!checked)} className={`mdi mdi-${checkboxIcon} ${checkboxIconClass}`} />
    </div>
  )
}
