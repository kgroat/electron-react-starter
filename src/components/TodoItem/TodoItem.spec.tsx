
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import TodoItem from './'
import { checkboxIcon, deleteIcon } from './TodoItem.scss'

describe('<TodoItem />', () => {
  describe('snapshots', () => {
    it('should render unchecked without a delete icon', () => {
      const tree = create(
        <TodoItem name='Get milk' />
      )
      expect(tree).toMatchSnapshot()
    })
    it('should render checked without a delete icon', () => {
      const tree = create(
        <TodoItem name='Get milk' checked />
      )
      expect(tree).toMatchSnapshot()
    })
    it('should render with a delete icon', () => {
      const tree = create(
        <TodoItem name='Get milk' onDelete={jest.fn()} />
      )
      expect(tree).toMatchSnapshot()
    })
  })

  describe('user clicking on checkbox', () => {
    it('should call onChange with true once', () => {
      const onChange = jest.fn()
      const element = shallow(
        <TodoItem name='Get milk' onChange={onChange} />
      )
      element.findWhere(el => el.hasClass(checkboxIcon)).simulate('click')
      expect(onChange).toHaveBeenCalledWith(true)
      expect(onChange).toHaveBeenCalledTimes(1)
    })
    it('should call onChange with false once', () => {
      const onChange = jest.fn()
      const element = shallow(
        <TodoItem name='Get milk' checked onChange={onChange} />
      )
      element.findWhere(el => el.hasClass(checkboxIcon)).simulate('click')
      expect(onChange).toHaveBeenCalledWith(false)
      expect(onChange).toHaveBeenCalledTimes(1)
    })
  })

  describe('user clicking on delete', () => {
    it('should call onDelete once', () => {
      const onDelete = jest.fn()
      const element = shallow(
        <TodoItem name='Get milk' onDelete={onDelete} />
      )
      element.findWhere(el => el.hasClass(deleteIcon)).simulate('click')
      expect(onDelete).toHaveBeenCalledTimes(1)
    })
  })
})
