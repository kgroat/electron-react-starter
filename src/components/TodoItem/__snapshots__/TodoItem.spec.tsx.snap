// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> snapshots should render checked without a delete icon 1`] = `
<div
  className="todoItem"
>
  <span
    className="title"
  >
    Get milk
  </span>
  <i
    className="mdi mdi-checkbox-marked-outline checkboxIcon"
    onClick={[Function]}
  />
</div>
`;

exports[`<TodoItem /> snapshots should render unchecked without a delete icon 1`] = `
<div
  className="todoItem"
>
  <span
    className="title"
  >
    Get milk
  </span>
  <i
    className="mdi mdi-checkbox-blank-outline checkboxIcon"
    onClick={[Function]}
  />
</div>
`;

exports[`<TodoItem /> snapshots should render with a delete icon 1`] = `
<div
  className="todoItem"
>
  <span
    className="title"
  >
    Get milk
  </span>
  <i
    className="mdi mdi-delete deleteIcon"
    onClick={[Function]}
  />
  <i
    className="mdi mdi-checkbox-blank-outline checkboxIcon"
    onClick={[Function]}
  />
</div>
`;
